src/sys/windows/selector.rs

```diff
@@ -449,129 +569,92 @@ impl SelectorInner {
-                if !sock_internal.is_pending_deletion() {
-                    sock_internal.update(&sock).unwrap();
-                }
+    unsafe fn update_sockets_events(&self) -> io::Result<()> {
+        let update_queue = &mut *self.update_queue.get();
+        loop {
+            let sock = match update_queue.pop_front() {
+                Some(sock) => sock,
+                None => break,
+            };
+            let mut sock_internal = sock.lock().unwrap();
+            if !sock_internal.is_pending_deletion() {
+                sock_internal.update(&sock).unwrap();
             }
         }
-        self._release_unused_afd();
+        self.afd_group.release_unused_afd();
         Ok(())
     }
 
-    fn update_sockets_events_if_polling(&self) -> io::Result<()> {
-        if self.active_poll_count.load(Ordering::SeqCst) > 0 {
+    unsafe fn update_sockets_events_if_polling(&self) -> io::Result<()> {
+        let active_poll_count = *self.active_poll_count.get();
+        if active_poll_count > 0 {
             return self.update_sockets_events();
         }
         Ok(())
     }
 
-    fn feed_events(&self, events: &mut Vec<Event>, iocp_events: &[CompletionStatus]) {
-        {
-            let mut update_queue = self.update_queue.lock().unwrap();
-            for iocp_event in iocp_events.iter() {
-                if iocp_event.overlapped() as usize == 0 {
-                    events.push(Event {
-                        flags: AFD_POLL_RECEIVE,
-                        data: iocp_event.token() as u64,
-                    });
-                    continue;
-                }
-                let sock_arc =
-                    unsafe { Arc::from_raw(iocp_event.overlapped() as *const Mutex<SockState>) };
-                let mut sock_guard = sock_arc.lock().unwrap();
-                match sock_guard.feed_event() {
-                    Some(e) => {
-                        events.push(e);
-                    }
-                    None => {}
-                }
-                if !sock_guard.is_pending_deletion() {
-                    update_queue.push_back(sock_arc.clone());
-                }
-            }
-        }
-        self._release_unused_afd();
+    unsafe fn add_socket_to_update_queue<S: SocketState>(&self, socket: &S) {
+        let sock_state = socket.get_sock_state().unwrap();
+        let update_queue = &mut *self.update_queue.get();
+        update_queue.push_back(sock_state);
     }
 
-    fn _acquire_afd(&self) -> io::Result<Arc<Afd>> {
-        let mut need_alloc = false;
-        {
-            let afd_group = self.afd_group.lock().unwrap();
-            if afd_group.len() == 0 {
-                need_alloc = true;
-            } else {
-                // + 1 reference in Vec
-                if Arc::strong_count(afd_group.last().unwrap()) >= POLL_GROUP__MAX_GROUP_SIZE + 1 {
-                    need_alloc = true;
+    // The ret value here means processed iocp_events not events itself.
```

Comment:
Can you replace this comment with it returns, i.e. Returns number of IOCP events processed (if that is correct).

Reply:
Also I don't see any path that doesn't process all `iocp_events`, so I don't really the point of the return value (I could be wrong though).

Reply:
Since Windows expected iocp events arrays are initialized with zeros, I think we can't just use the previous element.